jaeger-zipkin/src/main/java/com/uber/jaeger/propagation/b3/B3TextMapCodec.java

```diff
@@ -0,0 +1,116 @@
+import java.util.Map;
+import zipkin.internal.Util;
+
+import static com.uber.jaeger.SpanContext.contextFromString;
+
+/**
+ * This format is compatible with other trace libraries such as Brave, Wingtips, zipkin-js, etc.
+ *
+ * <p>Example usage:
+ * <pre>{@code
+ * b3Codec = new B3TextMapCodec();
+ * tracer = new Tracer.Builder(serviceName, reporter, sampler)
+ *                    .registerInjector(Format.Builtin.HTTP_HEADERS, b3Codec)
+ *                    .registerExtractor(Format.Builtin.HTTP_HEADERS, b3Codec)
+ *                    ...
+ * }</pre>
+ *
+ * <p>See <a href="http://zipkin.io/pages/instrumenting.html">Instrumenting a Library</a>
+ */
+public final class B3TextMapCodec implements Injector<TextMap>, Extractor<TextMap> {
+  static final String TRACE_ID_NAME = "X-B3-TraceId";
+  static final String SPAN_ID_NAME = "X-B3-SpanId";
+  static final String PARENT_SPAN_ID_NAME = "X-B3-ParentSpanId";
+  static final String SAMPLED_NAME = "X-B3-Sampled";
+  static final String FLAGS_NAME = "X-B3-Flags";
+  // NOTE: uber's flags aren't the same as B3/Finagle ones
+  static final byte SAMPLED_FLAG = 1;
+  static final byte DEBUG_FLAG = 2;
+
+  @Override
+  public void inject(SpanContext spanContext, TextMap carrier) {
+    carrier.put(TRACE_ID_NAME, Util.toLowerHex(spanContext.getTraceID()));
+    Long nullableParentId = nullableParentId(spanContext);
+    if (nullableParentId != null) {
+      carrier.put(PARENT_SPAN_ID_NAME, Util.toLowerHex(nullableParentId));
+    }
+    carrier.put(SPAN_ID_NAME, Util.toLowerHex(spanContext.getSpanID()));
+    carrier.put(SAMPLED_NAME, spanContext.isSampled() ? "1" : "0");
+    if (spanContext.isDebug()) {
+      carrier.put(FLAGS_NAME, "1");
+    }
+  }
+
+  @Override
+  public SpanContext extract(TextMap carrier) {
+    Long traceID = null;
+    Long spanID = null;
+    Long parentID = null;
+    byte flags = 0;
+    for (Map.Entry<String, String> entry : carrier) {
+      if (entry.getKey().equalsIgnoreCase(SAMPLED_NAME)) {
+        if (entry.getValue().equals("1") || entry.getValue().toLowerCase().equals("true")) {
+          flags |= SAMPLED_FLAG;
+        }
+      } else if (entry.getKey().equalsIgnoreCase(TRACE_ID_NAME)) {
+        traceID = Util.lowerHexToUnsignedLong(entry.getValue());
+      } else if (entry.getKey().equalsIgnoreCase(PARENT_SPAN_ID_NAME)) {
+        parentID = Util.lowerHexToUnsignedLong(entry.getValue());
+      } else if (entry.getKey().equalsIgnoreCase(SPAN_ID_NAME)) {
+        spanID = Util.lowerHexToUnsignedLong(entry.getValue());
+      } else if (entry.getKey().equalsIgnoreCase(FLAGS_NAME)) {
+        if (entry.getValue().equals("1")) {
+          flags |= DEBUG_FLAG;
+        }
+      }
+    }
+
+    // Uber's parent ID is not nullable
+    if (parentID == null) {
+      parentID = traceID;
+    }
+
+    if (traceID != null && spanID != null) {
+      return contextFromString(String.format("%x:%x:%x:%x", traceID, spanID, parentID, flags));
```

Comment:
this is kindof gross, but the there's no way to construct the context naturally from outside its package.